### ISSUE TITLE
Output closures appear to not be supported in OptiX

### ISSUE THREAD

**Issue by DeclanRussell** [imageworks/OpenShadingLanguage#1025]
### Problem

It appears that output closures are not supported when compiling OSL for OptiX.

**Expected behavior:**  
The shader should compile nicely into PTX ready for OptiX consumption.
**Actual behavior:**  
The code triggers and assert during the optimize_group method.

`C:\Some\Path\include\osl_pvt.h:275: failed assertion '! is_closure() && "Don't call this if it could be a closure"'`

### Steps to Reproduce

1. Attached is a modified version of testoptix that demonstrates the issue. You can run this with the following command line: testrender.exe -optix --debug2 -res 320 240 scene.xml out.exr
[OutputClosuresRepo.zip](https://github.com/imageworks/OpenShadingLanguage/files/3212065/OutputClosuresRepo.zip)

### Versions

* OSL branch/version: 1.11.0 -> master
* OS: windows
* C++ compiler: MSVC++ 14.0
* LLVM version: 8
* OIIO version: 2.1.0

**Comment by timgrant** [comment]
Declan, I have reproduced your issue and understand what's going wrong.

During ShaderGroup optimization, shader parameters are initialized differently depending on whether nor not OptiX is enabled. In the OptiX-specific code path, there are several places where the type of the parameter is queried. Performing certain queries on closure parameters is not permitted. In this case, checking if a closure parameter is a string triggers the assert.

The patch to avoid the assert is pretty small:

```diff --git a/src/liboslexec/llvm_instance.cpp b/src/liboslexec/llvm_instance.cpp
index d4a947c..af49828 100644
--- a/src/liboslexec/llvm_instance.cpp
+++ b/src/liboslexec/llvm_instance.cpp
@@ -497,7 +497,7 @@ BackendLLVM::llvm_assign_initial_value (const Symbol& sym, bool force)
         ll.op_branch (cond_val, no_userdata_block, after_userdata_block);
     }
 
-    if (use_optix() && ! sym.typespec().is_string()) {
+    if (use_optix() && ! sym.typespec().is_closure() && ! sym.typespec().is_string()) {
         ASSERT (! sym.has_init_ops() && "Init ops are not currently supported in OptiX");
 
         // If the call to osl_bind_interpolated_param returns 0, the default
@@ -510,7 +510,7 @@ BackendLLVM::llvm_assign_initial_value (const Symbol& sym, bool force)
 
         TypeDesc t = sym.typespec().simpletype();
         ll.op_memcpy (dst, src, t.size(), t.basesize());
-    } else if (use_optix()) {
+    } else if (use_optix() && ! sym.typespec().is_closure()) {
         // For convenience, we always pack string addresses into the groupdata
         // struct.
         int userdata_index = find_userdata_index (sym);
```
I added a light to your repro scene and the OptiX and CPU versions roughly match (within the limits of the OptiX renderer), but I'm not sure of the patch is sufficient for all use cases. But it's probably a good place to start.

**Comment by DeclanRussell** [comment]
Thanks for the patch Tim, I've only done some limited testing but it appears to work great and unblocks development for now. Would you be happy to create a PR to add these changes to master so I an close this issue?

**Comment by lgritz** [comment]
Who's gonna make a PR for this? Or should I?

**Comment by ee33** [comment]
Tim is making a PR.

**Comment by lgritz** [comment]
Merged #1029
